onnx/checker.cc

```diff
@@ -35,15 +43,138 @@ namespace checker {
     }                                         \
   } while (0)
 
-void check_value_info(const ValueInfoProto& value_info, const CheckerContext&) {
+bool check_name_syntax(const std::string& name) {
+  if (name.empty()) {
+    return false;
+  }
+
+  bool correct = true;
+
+  // Names should adhere to C identifier syntax.
+  auto iter = name.cbegin();
+
+  char c = *iter;
+  if (!(isalpha(c) || c == '_')) {
+    correct = false;
+  }
+
+  ++iter;
+
+  for (; iter < name.cend(); ++iter) {
+    c = *iter;
+    if (!(isalnum(c) || c == '_')) {
+      correct = false;
+    }
+  }
+
+  return correct;
+}
+
+bool check_domain_syntax(const std::string& name) {
+  if (name.empty()) {
+    return false;
+  }
+
+  bool correct = true;
+
+  // Names should adhere to C identifier syntax.
+  auto iter = name.cbegin();
+
+  char c = *iter;
+  if (!isalpha(c)) {
+    correct = false;
+  }
+
+  ++iter;
+
+  for (; iter < name.cend(); ++iter) {
+    c = *iter;
+    if (!(isalnum(c) || c == '.')) {
+      correct = false;
+    }
+  }
+
+  return correct;
+}
+
+#define enforce_c_identifier(proto, field)       \
+  do {                                           \
+    const auto& str = proto.field();             \
+    if (!check_name_syntax(str)) {               \
+      fail_check(                                \
+          ctx,                                   \
+          "'",                                   \
+          str,                                   \
+          "' is invalid for " #proto,            \
+          ".",                                   \
+          #field,                                \
+          ". It must use C identifier syntax."); \
+    }                                            \
+  } while (0)
+
+#define enforce_domain_name_rules(proto, field)     \
+  do {                                              \
+    const auto& str = proto.field();                \
+    if (!check_domain_syntax(str)) {                \
+      fail_check(                                   \
+          ctx,                                      \
+          "'",                                      \
+          str,                                      \
+          "' is invalid for " #proto,               \
+          ".",                                      \
+          #field,                                   \
+          ". It must be a valid DNS domain name."); \
+    }                                               \
+  } while (0)
+
+void check_value_info(const ValueInfoProto& value_info, CheckerContext& ctx) {
   enforce_non_empty_field(value_info, name);
   enforce_has_field(value_info, type);
   const auto value_case = value_info.type().value_case();
   switch (value_case) {
     case TypeProto::kTensorType: {
       const auto& type = value_info.type().tensor_type();
       enforce_has_field(type, elem_type);
-      enforce_has_field(type, shape);
+
+      if (type.has_shape()) {
+
+        for (auto& dim : type.shape().dim()) {
```

Comment:
We need to check `has_dim` before accessing it.

Reply:
Hmm, dim is a repeated field, so protoc doesn't generate a 'has_dim()' for it in C++.